src/types.rs

```diff
@@ -39,25 +88,42 @@ impl From<&HumanAddr> for HumanAddr {
     }
 }
 
+#[derive(Serialize, Deserialize, Clone, Default, Debug, PartialEq, JsonSchema)]
+pub struct CanonicalAddr(pub Base64);
+
+// CanonicalAddr is just a wrapper around Base64
+// TODO: make this a type alias???
+// pub type CanonicalAddr = Base64;
 impl CanonicalAddr {
     pub fn as_bytes(&self) -> &[u8] {
-        &self.0
+        self.0.as_bytes()
+    }
+    // decode will return the underlying bytes after decoding base64
+    pub fn decode(&self) -> Result<Vec<u8>> {
+        self.0.decode()
+    }
+    pub fn encode(data: &[u8]) -> Self {
+        CanonicalAddr(Base64::encode(data))
     }
     pub fn len(&self) -> usize {
         self.0.len()
     }
     pub fn is_empty(&self) -> bool {
         self.0.is_empty()
     }
+    /// this is meant to read in data from external APIs that pass in base64 data
+    /// it ensures proper utf8 string, but doesn't check base64 charset
+    pub fn from_external_base64(data: Vec<u8>) -> Result<Self> {
+        // TODO: make this unsafe to avoid checks?
+        // TODO: double-check base64 charset?
+        let s = String::from_utf8(data).context(Utf8StringErr {})?;
+        Ok(CanonicalAddr(Base64(s)))
```

Comment:
I think `CanonicalAddr` should store raw data binary. Base64 should be a transport format only, not a memory format. Or is this not possible with the automagick json encoding/decoding?

Reply:
It's not really possible. Or you can do it, but then it gets ugly if the dataformat (used by json-schema or other inspection tools) and the custom marshalling code don't line up. I have this issue with the cosmos-sdk where data types and json representations often have little to do with one another.

Also, my idea was that it is often fine just to use the base64 encoding to build keys - it is stable and canonical. Thus, `.as_bytes()`. If you really want the shorter version with the compute overhead, you can use `.decode()`

Basically, I define `CanonicalAddr` as "the base64 encoding of the blockchain-specific canonical binary representation of the address". And allow access to raw bytes as well. Is there a strong argument against it? My only thought was ~35% increase in storage size, but I think that is minor related to all the overhead in parsing otherwise